networkx/readwrite/text.py

```diff
@@ -72,6 +75,131 @@ def forest_str(graph, with_labels=True, sources=None, write=None, ascii_only=Fal
+    each unique node encountered. Non-tree edges are written to the right of
+    each node, and connection to a non-tree edge is indicated with an ellipsis.
+    This representation works best when the input graph is a forest, but any
+    graph can be represented.
+
+    Parameters
+    ----------
+    graph : nx.DiGraph | nx.Graph
+        Graph to represent
+
+    with_labels : bool
+        If True will use the "label" attribute of a node to display if it
+        exists otherwise it will use the node value itself. Defaults to True.
+
+    sources : List
+        Specifies which nodes to start traversal from. Note: nodes that are not
+        reachable from one of these sources may not be shown. If unspecified,
+        the minimal set of nodes needed to reach all others will be used.
+
+    write : callable
+        Function to use to write to, if None new lines are appended to
+        a list and returned. If set to the `print` function, lines will
+        be written to stdout as they are generated. If specified,
+        this function will return None. Defaults to None.
+
+    ascii_only : Boolean
+        If True only ASCII characters are used to construct the visualization
+
+    Returns
+    -------
+    str | None :
+        text representation of the graph
+
+    Example
+    -------
+    >>> graph = nx.balanced_tree(r=2, h=2, create_using=nx.DiGraph)
+    >>> print(nx.graph_str(graph))
+    ╙── 0
+        ├─╼ 1
+        │   ├─╼ 3
+        │   └─╼ 4
+        └─╼ 2
+            ├─╼ 5
+            └─╼ 6
+
+    >>> # A near tree with one non-tree edge
+    >>> graph.add_edge(5, 1)
+    >>> print(nx.graph_str(graph))
+    ╙── 0
+        ├─╼ 1 ╾ 5
+        │   ├─╼ 3
+        │   └─╼ 4
+        └─╼ 2
+            ├─╼ 5
+            │   └─╼  ...
+            └─╼ 6
+
+    >>> graph = nx.cycle_graph(5)
+    >>> print(nx.graph_str(graph))
+    ╙── 0
+        ├── 1
+        │   └── 2
+        │       └── 3
+        │           └── 4 ─ 0
+        └──  ...
+
+    >>> graph = nx.generators.barbell_graph(4, 2)
+    >>> print(nx.graph_str(graph))
+    ╙── 4
+        ├── 5
+        │   └── 6
+        │       ├── 7
+        │       │   ├── 8 ─ 6
+        │       │   │   └── 9 ─ 6, 7
+        │       │   └──  ...
+        │       └──  ...
+        └── 3
+            ├── 0
+            │   ├── 1 ─ 3
+            │   │   └── 2 ─ 0, 3
+            │   └──  ...
+            └──  ...
+
+    >>> graph = nx.complete_graph(5, create_using=nx.Graph)
+    >>> print(nx.graph_str(graph))
+    ╙── 0
+        ├── 1
+        │   ├── 2 ─ 0
+        │   │   ├── 3 ─ 0, 1
+        │   │   │   └── 4 ─ 0, 1, 2
+        │   │   └──  ...
+        │   └──  ...
+        └──  ...
+
+    >>> graph = nx.complete_graph(3, create_using=nx.DiGraph)
+    >>> print(nx.graph_str(graph))
+    ╙── 0 ╾ 1, 2
+        ├─╼ 1 ╾ 2
+        │   ├─╼ 2 ╾ 0
+        │   │   └─╼  ...
+        │   └─╼  ...
+        └─╼  ...
+    """
```

Comment:
Is there a reference that describes this notation for graphs in detail?

Reply:
I made this notation up, so I didn't follow any reference. I wouldn't be surprised if something like this does exist and has a detailed description, but I stumbled on it as a natural extension of the more obvious way to draw an "text forest". Let me see if I can summarize:

1. Given a set of source nodes (which can be specified, or automatically discovered via finding the CCs/SCCs and choosing one node with minimum degree from each), we traverse the graph in depth first order.

2. Each reachable node will be printed exactly once on it's own line.

3. Edges are indicated in one of three ways: 

    a. a parent "L-style" connection on the upper left. This corresponds to 
    b. a backref "<-style" connection shown directly on the right. For directed graphs, these are drawn for any incoming edges to a node that is not a parent edge. For undirected graphs, these are drawn for only the non-parent edges that have already been represented (The edges that have not been represented will be handled in the recursive case).
    c. a child "L-style" connection on the lower right. Drawing of the children are handled recursively.

4. The children of each node (wrt the directed DFS tree) are drawn underneath and to the left of it. In the case that a child node has already been drawn the connection is replaced with a "..." to indicate that there is one or more connections represented elsewhere. 
 

There might be a way to phrase this more completely / succinctly (or a reference to something that describes the same structure), but that's my stab at it.

Reply:
That text looks great!  Including the stuff at the top that we don't know a direct reference, but that there might be one. I guess you'd need to change the "I" to some sort of different voice like "This notation is made up and didn't follow any specific reference. Some document might exist somewhere for a basis for it, but this is a natural extension of one way to draw a text forest.  To summarize:"

Reply:
I've added a modified version of the text to the function.